src/main/java/org/sonarlint/intellij/analysis/SonarLintTask.java

```diff
@@ -0,0 +1,160 @@
+  }
+
+  public static SonarLintTask createBackground(IssueProcessor processor, SonarlintAnalyzer.SonarLintJob job) {
+    return new SonarLintTask(processor, job, true);
+  }
+
+  public static SonarLintTask createForeground(IssueProcessor processor, SonarlintAnalyzer.SonarLintJob job) {
+    return new SonarLintTask(processor, job, false);
+  }
+
+  @Override
+  public boolean shouldStartInBackground() {
+    return startInBackground;
+  }
+
+  private void stopRun() {
+    SonarLintStatus.get(job.module().getProject()).stopRun();
+  }
+
+  private static String getFileName(VirtualFile file) {
+    return file.getName();
+  }
+
+  @Override
+  public void run(ProgressIndicator indicator) {
+    SonarLintStatus status = SonarLintStatus.get(job.module().getProject());
+    try {
+      if (indicator.isCanceled() || status.isCanceled()) {
+        return;
+      }
+
+      indicator.setIndeterminate(true);
+      if (job.files().size() > 1) {
+        indicator.setText("Running SonarLint Analysis for " + job.files().size() + " files");
+      } else {
+        indicator.setText("Running SonarLint Analysis for '" + getFileName(job.files().iterator().next()) + "'");
+      }
+
+      final AccumulatorIssueListener listener = new AccumulatorIssueListener();
+      LOGGER.info(indicator.getText());
+
+      CancelMonitor monitor = new CancelMonitor(indicator, status, Thread.currentThread());
+      try {
+        monitor.start();
+        SonarLintAnalysisConfigurator.analyzeModule(job.module(), job.files(), listener);
+        indicator.startNonCancelableSection();
+      } finally {
+        monitor.stopMonitor();
+      }
+
+      //last chance to cancel (to avoid the possibility of having interrupt flag set)
+      if (indicator.isCanceled() || status.isCanceled()) {
+        return;
+      }
+
+      LOGGER.info("SonarLint analysis done");
+
+      indicator.setIndeterminate(false);
+      indicator.setFraction(.9);
+      indicator.setText("Creating SonarLint Issues: " + listener.getIssues().size());
+
+      processor.process(job, listener.getIssues());
+    } catch (RuntimeException e) {
+      // if cancelled, ignore any errors since they were most likely caused by the interrupt
+      if (!indicator.isCanceled() && !status.isCanceled()) {
+        throw e;
+      }
+    } finally {
+      stopRun();
+
+    }
+  }
+
+  private class CancelMonitor extends Thread {
+    private final ProgressIndicator indicator;
+    private final SonarLintStatus status;
+    private final Thread t;
+    private boolean stop = false;
+
+    public CancelMonitor(ProgressIndicator indicator, SonarLintStatus status, Thread t) {
+      this.indicator = indicator;
+      this.status = status;
+      this.t = t;
+      this.setName("SonarLint-cancel-monitor");
+      this.setDaemon(true);
+    }
+
+    public synchronized void stopMonitor() {
+      stop = true;
+    }
+
+    @Override
+    public void run() {
+      while (true) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Reduce the total number of break and continue statements in this loop to use at most one. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS135)